### PR TITLE
Mapping Event Flow Changes

### PR DESCRIPTION
**Pull request by mutaphysis**
These are actually three stacked branches. 
Feature are mostly complete, design is missing.

Feel free to take look at the code already :)

### CODE REVIEW

File: imports/both/api/MappingEvents/MappingEventSchema.ts [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-import { registerSchemaForI18n } from '../../i18n/i18n';
+import {registerSchemaForI18n} from '../../i18n/i18n';
+import {DraftEmailTemplateDataSchema} from './EmailTemplateSchema';
 import SimpleSchema from '../../lib/SimpleSchema';
-import ImageSchema from '../Images/ImageSchema';
+import {Meteor} from 'meteor/meteor';
+import {t} from 'ttag';
 import SearchResultFeatureSchema from '../shared/searchPlaces/SearchResultFeatureSchema';
-import { TrackingEventStatisticsSchema } from '../TrackingEvents/TrackingEventStatistics';
-import { eventStatusLabels } from './mappingEventStatus';
```
Comment (opyh) [inline imports/both/api/MappingEvents/MappingEventSchema.ts]: Did the `prettier` issue reappear?

File: imports/both/api/MappingEvents/buildInvitationEmailFromTemplate.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+type BodyTemplateData = {
+  appId: string,
+  appName: string,
+  eventId: string,
+  eventName: string,
+  organizationId: string,
+  organizationName: string,
+  token?: string
+  inviterUserName: string,
```
Comment (opyh) [inline imports/both/api/MappingEvents/buildInvitationEmailFromTemplate.ts]: Not sure, but maybe:

```suggestion
  hostUserName: string,
```

File: imports/client/pages/MappingEvents/MappingEventParticipantsPage.tsx [hunk 3]
```diff
@@ -189,6 +98,46 @@ const EventParticipantsListWithData = function (props: { eventId: string }) {
...
 
+const SendUpdateEmailForm = (props: { event: IMappingEvent }) => {
+  const {event} = props;
+
+  return <EmailTemplateForm
+    model={event.updateEmailData}
+    schema={UpdateEmailDataSchema}
+    onSubmit={(data) => {
+      const sendPromise = sendMappingEventUpdateEmail(String(event._id), 'updateEmailData', data);
+      sendPromise
```
Comment (opyh) [inline imports/client/pages/MappingEvents/MappingEventParticipantsPage.tsx]: -> `async`/`await`?

File: imports/client/pages/MappingEvents/components/InviteLinkSection.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    }
+  }
+
+  section {
+    canvas {
+      height: calc(196px) !important;
+      width: calc(196px) !important;
+      object-fit: contain;
+      image-rendering: pixelated;
```
Comment (opyh) [inline imports/client/pages/MappingEvents/components/InviteLinkSection.tsx]: 😀

File: imports/client/components/Forms/EditForm.tsx [hunk 5]
```diff
@@ -24,6 +24,7 @@ interface IProps<T, D extends ICollectionDefinition<T>> {
   /// If `_id` is not set, the form will insert a new document.
   _id?: string;
   mode: 'insert' | 'update',
+  submitText?: string,
```
Comment (opyh) [inline imports/client/components/Forms/EditForm.tsx]:
```suggestion
  submitButtonCaption?: string,
```

File: imports/both/api/MappingEvents/IMappingEvent.ts [hunk 6]
```diff
@@ -24,8 +42,14 @@ type IMappingEvent = {
   photos?: IImage[];
   status: EventStatusEnum;
   statistics: IMappingEventStatistics;
+  invitationEmailData?: DraftEmailTemplateData;
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/both/api/MappingEvents/IMappingEvent.ts]: It's not the template, it's the data to fill a template...

File: imports/both/api/MappingEvents/EmailTemplateSchema.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  subject: {
+    type: String,
+    optional: true,
+    label: t`Subject`,
+  },
+  message: {
+    type: String,
+    optional: true,
+    label: t`Message`,
+    uniforms: {
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/both/api/MappingEvents/EmailTemplateSchema.ts]: We have this pattern already in plenty of places, so would be a bit more work to fix all the occurences entirely, but I can move this into the frontend without problems.

File: imports/both/api/MappingEvents/MappingEventSchema.ts (hunk 1 above)
Comment (mutaphysis) [reply imports/both/api/MappingEvents/MappingEventSchema.ts]: Rather different rules when reordering imports.

File: imports/client/pages/MappingEvents/MappingEventParticipantsPage.tsx (hunk 3 above)
Comment (mutaphysis) [reply imports/client/pages/MappingEvents/MappingEventParticipantsPage.tsx]: The awaiting happens inside the EmailTemplateForm

File: imports/client/components/Forms/EditForm.tsx (hunk 5 above)
Comment (mutaphysis) [reply imports/client/components/Forms/EditForm.tsx]: Yep, way better!

File: imports/both/api/MappingEvents/DraftEmailTemplateData.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+export type TemplateTypeEnum = 'informal' | 'formal';
+
+export const emailTemplateTypes = Object.freeze([
+  {label: t`Informal template`, value: 'informal'},
+  {label: t`Formal template`, value: 'formal'},
+]) as ReadonlyArray<{ label: string; value: TemplateTypeEnum }>;
+
+export type DraftEmailTemplateData = {
+  template?: TemplateTypeEnum,
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/both/api/MappingEvents/DraftEmailTemplateData.ts]: Better, will change.

File: imports/both/api/MappingEvents/buildInvitationEmailFromTemplate.ts (hunk 2 above)
Comment (mutaphysis) [reply imports/both/api/MappingEvents/buildInvitationEmailFromTemplate.ts]: Not necessarily true. The person that sends out the invite is might not be the host. But I wasn't happy with the name either.

File: imports/both/api/MappingEvents/EmailTemplateSchema.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/both/api/MappingEvents/EmailTemplateSchema.ts]: Done

File: imports/both/api/MappingEvents/MappingEventSchema.ts (hunk 1 above)
Comment (opyh) [reply imports/both/api/MappingEvents/MappingEventSchema.ts]: Sorry, that wasn't clear enough - I meant the whitespacing / indentation differences that suddenly reappeared… when we tested last time, it seemed to indent the same way on my and your machine, now it appears it doesn't anymore. Or do I overlook something?

File: imports/both/api/MappingEvents/buildInvitationEmailFromTemplate.ts (hunk 2 above)
Comment (opyh) [reply imports/both/api/MappingEvents/buildInvitationEmailFromTemplate.ts]: I've looked other synonyms up and didn't find anything better, so let's keep it as at least it's 100% clear what it means :)

File: imports/both/api/MappingEvents/MappingEventSchema.ts (hunk 1 above)
Comment (mutaphysis) [reply imports/both/api/MappingEvents/MappingEventSchema.ts]: That was a different project :)

File: imports/client/pages/MappingEvents/MappingEventParticipantsPage.tsx (hunk 3 above)
Comment (opyh) [reply imports/client/pages/MappingEvents/MappingEventParticipantsPage.tsx]: I meant the `.then`/`.catch` inside this function could be written in async/await style. I can personally visually parse both styles pretty quickly as I've worked with promises for years, but heard feedback from other coders that they had problems with this even if they know how the concept works. Just a suggestion though, I'm fine with leaving it as it is :)

**Comment by mutaphysis** [comment]
Most planned features are done, I will add new features in different branches from now on.

File: imports/server/json-api/TrackingEvents/trackingStatistics.ts [hunk 9]
```diff
@@ -59,8 +72,47 @@ const eventStatisticsConfigurations: EventStatisticsConfigurations = {
...
+            })
+          : null;
+        let existingParticipantId: string | undefined = undefined;
+        if (participantByToken) {
+          // token before email, when both are there...
+          existingParticipantId = participantByToken._id;
+        } else if (participantByEmail) {
+          existingParticipantId = participantByEmail._id;
+        }
+        if (existingParticipantId) {
```
Comment (opyh) [inline imports/server/json-api/TrackingEvents/trackingStatistics.ts]: Will clarify with Holger how to deal with this behavior - for now it's okay :)

**Pull request merged**